app/src/main/kotlin/compiler/intermediate_form/FunctionDependenciesAnalyzer.kt

```diff
@@ -24,18 +24,39 @@ object FunctionDependenciesAnalyzer {
             return uniqueIdentifiers[function]!!.value
         }
 
-        fun analyze(node: Any, pathSoFar: String?) {
+        fun analyze(node: Any, pathSoFar: String? = null) {
             when (node) {
                 is Program.Global.FunctionDefinition -> { analyze(node.function, pathSoFar) }
                 is Statement.FunctionDefinition -> { analyze(node.function, pathSoFar) }
                 is Function -> {
                     val newPrefix = nameFunction(node, pathSoFar)
-                    node.body.forEach { analyze(it, newPrefix) }
+                    var blockNumber = -1
+                    for (statement in node.body) {
+                        when (statement) {
+                            is Statement.Block -> {
+                                blockNumber++
+                                statement.block.forEach { analyze(it, newPrefix + blockNumber) }
+                                analyze(statement, newPrefix + blockNumber)
+                            }
+                            is Statement.Conditional -> {
+                                blockNumber++
+                                statement.actionWhenTrue.forEach { analyze(it, newPrefix + blockNumber) }
+                                blockNumber++
+                                statement.actionWhenFalse?.forEach { analyze(it, newPrefix + blockNumber) }
+                            }
+                            is Statement.Loop -> {
+                                blockNumber++
```

Comment:
Nit: There is a lot of code repetition. Maybe you could create a nested function for analyzing lists, sth like:

`fun analyzeList(nodes: List<Statement>?) =nodes?.forEach {analyze(it, newPrefix + blockNumber++)}`
(and start blockNumber from 0 then)